d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java

```diff
@@ -627,6 +671,55 @@ public void setDelayedExecution(long delayedExecution)
     _delayedExecution = delayedExecution;
   }
 
+  @Override
+  public Map<URI, TrackerClient> getClientsSubset(String serviceName,
+                                                  ServiceProperties serviceProperties,
+                                                  int partitionId,
+                                                  Map<URI, TrackerClient> potentialClients)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, serviceProperties, partitionId);
+    if (subsettingStrategy != null)
+    {
+      if (subsettingStrategy.isSubsetChanged(_version.get()) ||
+          !_weightedSubsetsCache.containsKey(serviceName) ||
+          !_weightedSubsetsCache.get(serviceName).containsKey(partitionId))
```

Comment:
I would prefer to decouple subsetting strategy from the simpleLoadBalancerState caching logic. First, I would like to keep subsetting strategy as a general interface, as to pick a subset from a collection of items. This is also how the D2 Hash Ring works, as the interface suggests 

public interface Ring<T>
{
  /**
   * Pick an object in the ring given a specified key
   *
   * @param key   The value used to pick an object from the hash ring
   * @return An object in the ring
   */
  @Nullable
  T get(int key);
}

Second, the caching logic also requires setting subset weight on TrackerClient, which I think is specific to SimpleLoadBalancerState.

As for exposing isSubsetChanged as a public interface, we need this because we don't want to set TrackerClient subset weight over and over again when the weight map doesn't change. If the subsetting strategy only returns the subset, the SimpleLoadBalancer has no idea whether the subset is changed from previous one.

Reply:
Alright, I'm convinced, when there're multiple implementations of SubsettingStrategy, the cache management logic should be the same.